### DISCUSSION TITLE
New type of angular spread source: read in distribution from file

### DISCUSSION THREAD

**Discussion by mcoolman** [nrc-cnrc/EGSnrc#771] [Ideas]
A colleague has suggested that I should post my self implemented source here. I implemented a source as a new type of angular spread source, that can read in a distribution from file. The distribution is read in the same way as spectra and stored in an alias_table object. 

The file with the distribution has a very similar formatting as a spectrum file. I attached an example. Then the program samples an angle from the distribution and applies it. You can choose the direction of the spread (x, y, or both). Therefore I added two methods in the `egs_vector.h` file:
```
    void rotate_in_xz(EGS_Float cos_t, EGS_Float sin_t) {

            x = sin_t;
            z = cos_t;

            x = x/sqrt(1+y*y);
            y = y/sqrt(1+y*y);
            z = z/sqrt(1+y*y);

    };

    void rotate_in_yz(EGS_Float cos_t, EGS_Float sin_t) {

            y = sin_t;
            z = cos_t;

            x = x/sqrt(1+x*x);
            y = y/sqrt(1+x*x);
            z = z/sqrt(1+x*x);

    };
```

I attached the implementation files (I was not able to attach them as `.cpp`, `.h` and `Makefile` files, so I attached them as `.txt`). It is just a first attempt and works for my goals. I'm not fully sure that it works for every case. If you are interested you can have a look. I am happy about any ideas, comments or improvements.

Greetings
Malu

---

Use in `.egsinp`:

```
    :start source definition:
       :start source:
           # define a source, on which you want to apply the spread.
           name = any source
       :stop source

       :start source:
            library = egs_angular_spread_distribution_source
            name = my_source
            source name = any_source
            distribution file = #path to your distribution file
            direction = 0 # 0 = x, 1 = y, 2 = both
       :stop source:

    
       simulation source = my_source

    :stop source definition:
```

[80kV_body_prob_distribution.txt](https://github.com/nrc-cnrc/EGSnrc/files/7250344/80kV_body_prob_distribution.txt)
[egs_angular_spread_distribution_source_cpp.txt](https://github.com/nrc-cnrc/EGSnrc/files/7250370/egs_angular_spread_distribution_source_cpp.txt)
[egs_angular_spread_distribution_source_h.txt](https://github.com/nrc-cnrc/EGSnrc/files/7250371/egs_angular_spread_distribution_source_h.txt)
[Makefile.txt](https://github.com/nrc-cnrc/EGSnrc/files/7250372/Makefile.txt)

**Comment by ftessier** [comment]
@mcoolman, thank you for this contribution. If you can, please post your changes as a Pull Request (PR), so we can discuss, edit and eventually merge your code into the EGSnrc project. We can create the pull request if you prefer, but it is best if you do it, to ensure that authors gets the credit they deserve, as your name would be recorded in the commits.

To draft a PR, you will need to fork the EGSnrc repo in your area (@mcoolman), create a new branch, modify the files (or add new files), then create a PR. See [instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork) for more details, or let me know if I can assist you in the process.
